MAKE FASTLANE FAST AGAIN

neonichu opened the pull request:
Turns out `fastlane` can be kinda slow if the user has a bunch of gems installed:

``` bash
$ time fastlane -v
real    1m36.253s
user    1m28.818s
sys 0m2.793s
```

with the changes in this branch:

``` bash
$ time fastlane -v
real    0m2.277s
user    0m1.704s
sys 0m0.295s
```

The approach is kinda hacky, it works by shelling out to `bundle exec fastlane` and generates a `Gemfile` on the fly if there isn't one already.

cc @KrauseFx

KrauseFx commented:
But is it fast enough?

neonichu commented:
🏃

KrauseFx commented on fastlane/bin/fastlane:
```diff
@@ -1,5 +1,17 @@
 
+unless ENV['BUNDLE_BIN_PATH']
+  has_gemfile = File.exist?('Gemfile')
+  File.open('Gemfile', 'w') { |file| file.write("gem 'fastlane'") } unless has_gemfile
```
Could we also create this file in a temporary directory, and use it from there? This way, users won't see the Gemfile when they look at their git client while fastlane is running

KrauseFx replied:
Alternatively fastlane could delete temporary Gemfiles itself once it is launched, this way the temporary file is there only during dependency resolving

yanndupuy replied:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  dupuyyann@gmail.com
    Domain zyapps.com has exceeded the max emails per hour (450/400 (112%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext8.iad.github.net ([192.30.252.199]:36369 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by serveur.ex2-p10.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1bEKsb-002lO1-S7
	for yann.dupuy@zyapps.com; Sat, 18 Jun 2016 20:22:06 +0200
Date: Sat, 18 Jun 2016 11:21:20 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1466274080;
	bh=KXs9i16XFZKP9rGoMBzTMEh9+3TbGDROOUanjEPEkPE=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=oT9Z9IqUAga4KR40k/FRwSh36NZAsQeRWMK2ZJ1tKGC2atqj2U89HCV9obDUTfPOI
	 OqvwIgmw1os76lvblMKdBucPbXzoEmWVurh5WWWGV6O1ycn52BY4/xVkXkIQP04wd7
	 fylWKeTX7RtJk/2uf/Q9yWBGXlRX/jgt+dlvqZFY=
From: Felix Krause <notifications@github.com>
Reply-To: fastlane/fastlane <reply@reply.github.com>
To: fastlane/fastlane <fastlane@noreply.github.com>
Cc:
Message-ID: <fastlane/fastlane/pull/5121/r67604615@github.com>
In-Reply-To: <fastlane/fastlane/pull/5121@github.com>
References: <fastlane/fastlane/pull/5121@github.com>
Subject: Re: [fastlane/fastlane] MAKE FASTLANE FAST AGAIN (#5121)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_576591208967a_1b0b3f9b9c98129c1611d1";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: KrauseFx
X-GitHub-Recipient: yanndupuy
List-ID: fastlane/fastlane <fastlane.fastlane.github.com>
List-Archive: https://github.com/fastlane/fastlane
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000df0dc6ef2b2658cdb94cb10c1f1e9b3e5019e649e930a92cf00000001137d532092a170ce04079087@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AA3w3NAxXWl9E5pGJAg_KtPl90zn96Wpks5qNDcggaJpZM4I0GOk>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: yann.dupuy@zyapps.com
X-Spam-Status: No, score=-4.4
X-Spam-Score: -43
X-Spam-Bar: ----
X-Ham-Report: Spam detection software, running on the system "serveur.ex2-p10.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  > @@ -1,5 +1,17 @@ > #!/usr/bin/env ruby > $LOAD_PATH.push
    File.expand_path("../../lib", __FILE__) > > +unless ENV['BUNDLE_BIN_PATH']
    > + has_gemfile = File.exist?('Gemfile') > + File.open('Gemfile', 'w') {
   |file| file.write("gem 'fastlane'") } unless has_gemfile [...] 
 
 Content analysis details:   (-4.4 points, 5.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -5.0 RCVD_IN_DNSWL_HI       RBL: Sender listed at http://www.dnswl.org/, high
                             trust
                             [192.30.252.199 listed in list.dnswl.org]
 -0.0 RCVD_IN_MSPIKE_H2      RBL: Average reputation (+2)
                             [192.30.252.199 listed in wl.mailspike.net]
 -0.0 SPF_PASS               SPF: sender matches SPF record
  0.0 HTML_MESSAGE           BODY: HTML included in message
  0.7 HTML_IMAGE_ONLY_20     BODY: HTML: images with 1600-2000 bytes of words
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
X-Spam-Flag: NO


----==_mimepart_576591208967a_1b0b3f9b9c98129c1611d1
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

> @@ -1,5 +1,17 @@
>  #!/usr/bin/env ruby
>  $LOAD_PATH.push File.expand_path("../../lib", __FILE__)
>  
> +unless ENV['BUNDLE_BIN_PATH']
> +  has_gemfile = File.exist?('Gemfile')
> +  File.open('Gemfile', 'w') { |file| file.write("gem 'fastlane'") } unless has_gemfile

Could we also create this file in a temporary directory, and use it from there? This way, users won't see the Gemfile when they look at their git client while fastlane is running

---
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/fastlane/fastlane/pull/5121/files/24998c0819c6c514e1543f6c371680a066ce2137#r67604615
----==_mimepart_576591208967a_1b0b3f9b9c98129c1611d1
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>In <a href="https://github.com/fastlane/fastlane/pull/5121#discussion_r67604615">fastlane/bin/fastlane</a>:</p>
<pre style='color:#555'>&gt; @@ -1,5 +1,17 @@
&gt;  #!/usr/bin/env ruby
&gt;  $LOAD_PATH.push File.expand_path(&quot;../../lib&quot;, __FILE__)
&gt;  
&gt; +unless ENV[&#39;BUNDLE_BIN_PATH&#39;]
&gt; +  has_gemfile = File.exist?(&#39;Gemfile&#39;)
&gt; +  File.open(&#39;Gemfile&#39;, &#39;w&#39;) { |file| file.write(&quot;gem &#39;fastlane&#39;&quot;) } unless has_gemfile
</pre>
<p>Could we also create this file in a temporary directory, and use it from there? This way, users won't see the Gemfile when they look at their git client while fastlane is running</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/fastlane/fastlane/pull/5121/files/24998c0819c6c514e1543f6c371680a066ce2137#r67604615">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe/AA3w3NAxXWl9E5pGJAg_KtPl90zn96Wpks5qNDcggaJpZM4I0GOk">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AA3w3FhArhJdEzhkPgdVP5X8Blg_Tj2Kks5qNDcggaJpZM4I0GOk.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/fastlane/fastlane/pull/5121/files/24998c0819c6c514e1543f6c371680a066ce2137#r67604615"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_576591208967a_1b0b3f9b9c98129c1611d1--

yanndupuy replied:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  dupuyyann@gmail.com
    Domain zyapps.com has exceeded the max emails per hour (471/400 (117%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o3.sgmail.github.com ([192.254.112.98]:51818)
	by serveur.ex2-p10.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.87)
	(envelope-from <bounces+848413-8f06-yann.dupuy=zyapps.com@sgmail.github.com>)
	id 1bEKtR-002ld9-7Z
	for yann.dupuy@zyapps.com; Sat, 18 Jun 2016 20:22:58 +0200
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=AAehinwnZp0nFyXqgumMAr5LzdQ=; b=JaToEC8MT3Am5dcJ
	+INA2bsUkK9X1jGeRS7jAqTecuS28P3RusCyxU2VihOr2qxFUH7mE0klA7AwLHKm
	TiMYKVItUjzZ30BUCvzgvzGassdt9NDy00mCVZDf2CnOwJ0AVymuAAAlUbg5uFk2
	UKXut8PnRl5p1cNC6fkT/oHxCWo=
Received: by filter0438p1mdw1.sendgrid.net with SMTP id filter0438p1mdw1.21394.576591462B
        2016-06-18 18:21:58.972573358 +0000 UTC
Received: from github-smtp2b-ext-cp1-prd.iad.github.net (github-smtp2b-ext-cp1-prd.iad.github.net [192.30.253.17])
	by ismtpd0006p1iad1.sendgrid.net (SG) with ESMTP id i_mY5WMXTMq5I9Rg6KWl7w
	for <yann.dupuy@zyapps.com>; Sat, 18 Jun 2016 18:21:58.807 +0000 (UTC)
Date: Sat, 18 Jun 2016 11:21:58 -0700
From: Felix Krause <notifications@github.com>
Reply-To: fastlane/fastlane <reply@reply.github.com>
To: fastlane/fastlane <fastlane@noreply.github.com>
Cc:
Message-ID: <fastlane/fastlane/pull/5121/r67604623@github.com>
In-Reply-To: <fastlane/fastlane/pull/5121@github.com>
References: <fastlane/fastlane/pull/5121@github.com>
Subject: Re: [fastlane/fastlane] MAKE FASTLANE FAST AGAIN (#5121)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_57659146a460a_537d3fa92afa12bc121297f";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: KrauseFx
X-GitHub-Recipient: yanndupuy
List-ID: fastlane/fastlane <fastlane.fastlane.github.com>
List-Archive: https://github.com/fastlane/fastlane
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000df0dc06264436a4ab1b04c9bec6029c0a399b11492ea792cf00000001137d534692a170ce0407908f@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AA3w3BF-hQF7JgyEXt_r8H0YnStjsXKdks5qNDdGgaJpZM4I0GOk>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: yann.dupuy@zyapps.com
X-SG-EID: /bDRkhSo/GH2gzZXMo5AKhbuTgLDPkexBjh/BYa140G67x3bv6g5vlEQ4OMEylZF6ff+ahNlryKtie
 KrsaMZ/KGCIWy6z18l/dB3Cti+OWHNJPSwNh3LdJeG8ieWRdm55uLXu9LmIvQJwM0e256fepuGVuU0
 NILz/bsRgesUjePUGK9MENNhDFB7CrXS7kb0ytZArA1tREq4cM/1JFsB2JM13hjuYx+whyArtCjTmc
 Q=
X-Spam-Status: No, score=-3.1
X-Spam-Score: -30
X-Spam-Bar: ---
X-Ham-Report: Spam detection software, running on the system "serveur.ex2-p10.com",
 has NOT identified this incoming email as spam.  The original
 message has been attached to this so you can view it or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  > @@ -1,5 +1,17 @@ > #!/usr/bin/env ruby > $LOAD_PATH.push
    File.expand_path("../../lib", __FILE__) > > +unless ENV['BUNDLE_BIN_PATH']
    > + has_gemfile = File.exist?('Gemfile') > + File.open('Gemfile', 'w') {
   |file| file.write("gem 'fastlane'") } unless has_gemfile [...] 
 
 Content analysis details:   (-3.1 points, 5.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -2.3 RCVD_IN_DNSWL_MED      RBL: Sender listed at http://www.dnswl.org/, medium
                              trust
                             [192.254.112.98 listed in list.dnswl.org]
 -0.0 RCVD_IN_MSPIKE_H3      RBL: Good reputation (+3)
                             [192.254.112.98 listed in wl.mailspike.net]
 -1.4 RP_MATCHES_RCVD        Envelope sender domain matches handover relay domain
  0.0 HTML_MESSAGE           BODY: HTML included in message
  0.7 HTML_IMAGE_ONLY_20     BODY: HTML: images with 1600-2000 bytes of words
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.0 RCVD_IN_MSPIKE_WL      Mailspike good senders
X-Spam-Flag: NO

----==_mimepart_57659146a460a_537d3fa92afa12bc121297f
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

> @@ -1,5 +1,17 @@
>  #!/usr/bin/env ruby
>  $LOAD_PATH.push File.expand_path("../../lib", __FILE__)
>  
> +unless ENV['BUNDLE_BIN_PATH']
> +  has_gemfile = File.exist?('Gemfile')
> +  File.open('Gemfile', 'w') { |file| file.write("gem 'fastlane'") } unless has_gemfile

Alternatively fastlane could delete temporary Gemfiles itself once it is launched, this way the temporary file is there only during dependency resolving

---
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/fastlane/fastlane/pull/5121/files/24998c0819c6c514e1543f6c371680a066ce2137#r67604623
----==_mimepart_57659146a460a_537d3fa92afa12bc121297f
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>In <a href="https://github.com/fastlane/fastlane/pull/5121#discussion_r67604623">fastlane/bin/fastlane</a>:</p>
<pre style='color:#555'>&gt; @@ -1,5 +1,17 @@
&gt;  #!/usr/bin/env ruby
&gt;  $LOAD_PATH.push File.expand_path(&quot;../../lib&quot;, __FILE__)
&gt;  
&gt; +unless ENV[&#39;BUNDLE_BIN_PATH&#39;]
&gt; +  has_gemfile = File.exist?(&#39;Gemfile&#39;)
&gt; +  File.open(&#39;Gemfile&#39;, &#39;w&#39;) { |file| file.write(&quot;gem &#39;fastlane&#39;&quot;) } unless has_gemfile
</pre>
<p>Alternatively fastlane could delete temporary Gemfiles itself once it is launched, this way the temporary file is there only during dependency resolving</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/fastlane/fastlane/pull/5121/files/24998c0819c6c514e1543f6c371680a066ce2137#r67604623">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe/AA3w3BF-hQF7JgyEXt_r8H0YnStjsXKdks5qNDdGgaJpZM4I0GOk">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AA3w3Ax04JOPK1T0cfX6lQCkhsPRhWqDks5qNDdGgaJpZM4I0GOk.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/fastlane/fastlane/pull/5121/files/24998c0819c6c514e1543f6c371680a066ce2137#r67604623"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_57659146a460a_537d3fa92afa12bc121297f--

KrauseFx commented:
Let me know when it's ready for a review 👍

neonichu committed:
Always be bundling

- Shell out to `bundle exec fastlane` if not already bundling
- Generate a `Gemfile` on the fly if necessary

neonichu commented:
🚀🚀🚀

KrauseFx commented:
`danger` broke the build again. Could you rebase?

KrauseFx commented:
Ah, because you squashed

KrauseFx commented:
Thanks for updating the PR 👍

neonichu commented:
:rocket: :rocket: :rocket:

KrauseFx commented on fastlane/bin/🚀:
```diff
@@ -1,5 +1,2 @@
-
-require "fastlane/cli_tools_distributor"
-Fastlane::CLIToolsDistributor.take_off
+load(File.expand_path('../fastlane', __FILE__))
```
I'll try that, no idea if that works 👍

neonichu replied:
It does :)

KrauseFx commented on fastlane/bin/fastlane:
```diff
@@ -1,5 +1,23 @@
+  has_gemfile = File.exist?('Gemfile')
+
+  require 'tmpdir'
+  temp_gemfile = "#{Dir.tmpdir}/Gemfile"
```
Should be `File.join(Dir.tmpdir, "Gemfile")`

KrauseFx commented on fastlane/bin/fastlane:
```diff
@@ -1,5 +1,23 @@
+
+  require 'tmpdir'
+  temp_gemfile = "#{Dir.tmpdir}/Gemfile"
+  File.open(temp_gemfile, 'w') { |file| file.write("gem 'fastlane'") } unless has_gemfile
```
Should be `File.write(temp_gemfile, "gem 'fastlane'")`

KrauseFx commented on fastlane/bin/fastlane:
```diff
@@ -1,5 +1,23 @@
+  temp_gemfile = "#{Dir.tmpdir}/Gemfile"
+  File.open(temp_gemfile, 'w') { |file| file.write("gem 'fastlane'") } unless has_gemfile
+
+  system("BUNDLE_GEMFILE=#{temp_gemfile} bundle exec fastlane #{ARGV.join(' ')}")
```
I'll have to test this 👍

KrauseFx commented:
Added some comments

neonichu committed:
Adressed some feedback

neonichu committed:
make fastlane work again

segiddins commented on fastlane/bin/fastlane:
```diff
@@ -1,5 +1,25 @@
 $LOAD_PATH.push File.expand_path("../../lib", __FILE__)
 
+unless ENV['BUNDLE_BIN_PATH']
+  has_gemfile = File.exist?('Gemfile') && File.read('Gemfile').include?('fastlane')
```
please use `Bundler::SharedHelper.in_bundle?` since users can use different gemfile names than just `Gemfile`

neonichu committed:
Remove automatic `Gemfile` generation

This was too brittle, so instead we're shelling out to `bundle exec`
only if there is already an existing `Gemfile` which includes
`fastlane`.

KrauseFx commented:
This looks great from my side, but needs some testing before we should ship this 🚀

KrauseFx commented:
What happens when:
- You have a `Gemfile` with `fastlane` as dependency
- You clone the repo on a new machine where you didn't run `bundle install` yet
- Just running `fastlane [lane]` would work, as it's using any version of `fastlane`
- With this PR merged this will fail if you have a different version installed, telling the user to run `bundle install`, right? 

Just making sure I didn't forget anything, we might break setups with this change, if people don't use `bundle exec`

KrauseFx commented:
Thanks @neonichu for the effort, but as you know we solved this by detecting the speed and telling the user how to create a Gemfile 👍

neonichu commented:
"solved" 😛
